File: project/iOS/imitate/Manager/AppLogger.swift
import OSLog

// MARK: - Log Destination (for future Firebase integration)

protocol LogDestination {
    func log(level: AppLogger.Level, category: AppLogger.Category, message: String)
}

// MARK: - AppLogger

final class AppLogger {

    enum Level {
        case debug, info, warning, error
    }

    enum Category: String {
        case screen  = "Screen"
        case action  = "Action"
        case channel = "Channel"
    }

    static let shared = AppLogger()

    var destinations: [LogDestination] = []

    private let subsystem = Bundle.main.bundleIdentifier ?? "com.imitate.app"

    private lazy var screenLogger  = Logger(subsystem: subsystem, category: Category.screen.rawValue)
    private lazy var actionLogger  = Logger(subsystem: subsystem, category: Category.action.rawValue)
    private lazy var channelLogger = Logger(subsystem: subsystem, category: Category.channel.rawValue)

    private init() {}

    // MARK: - Screen

    func screenAppeared(file: String = #fileID) {
        let message = "【iOS】[\(file)] appeared"
        screenLogger.debug("\(message, privacy: .public)")
        forward(level: .debug, category: .screen, message: message)
    }

    // MARK: - Action

    func userAction(_ action: String, on screen: String = #fileID) {
        let message = "【iOS】[\(screen)] '\(action)' action"
        actionLogger.debug("\(message, privacy: .public)")
        forward(level: .debug, category: .action, message: message)
    }

    // MARK: - Channel

    func channelRequest(_ method: String) {
        let message = "【iOS】[\(method)] channel requesting"
        channelLogger.debug("\(message, privacy: .public)")
        forward(level: .debug, category: .channel, message: message)
    }

    func channelSuccess(_ method: String) {
        let message = "【iOS】[\(method)] channel success"
        channelLogger.debug("\(message, privacy: .public)")
        forward(level: .debug, category: .channel, message: message)
    }

    func channelFailure(_ method: String, error: String) {
        let message = "【iOS】[\(method)] channel failure: \(error)"
        channelLogger.debug("\(message, privacy: .public)")
        forward(level: .debug, category: .channel, message: message)
    }

    // MARK: - Private

    // TODO: Firebase導入時に FirebaseLogDestination を実装し destinations に追加する
    private func forward(level: Level, category: Category, message: String) {
        destinations.forEach { $0.log(level: level, category: category, message: message) }
    }
}

File: project/Android/app/src/main/java/com/example/imitate_android/manager/AppLogger.kt
package com.example.imitate_android.manager

import android.util.Log
import com.example.imitate_android.BuildConfig

interface LogDestination {
    fun log(level: AppLogger.Level, category: AppLogger.Category, message: String)
}

object AppLogger {

    enum class Level {
        DEBUG, INFO, WARNING, ERROR
    }

    enum class Category(val tag: String) {
        SCREEN("Screen"),
        ACTION("Action"),
        CHANNEL("Channel")
    }

    var destinations: MutableList<LogDestination> = mutableListOf()

    // MARK: - Screen

    fun screenAppeared(screen: String) {
        val message = "【Android】[$screen] appeared"
        debugLog(Category.SCREEN, message)
        forward(Level.DEBUG, Category.SCREEN, message)
    }

    // MARK: - Action

    fun userAction(action: String, screen: String) {
        val message = "【Android】[$screen] '$action' action"
        debugLog(Category.ACTION, message)
        forward(Level.DEBUG, Category.ACTION, message)
    }

    // MARK: - Channel

    fun channelRequest(method: String) {
        val message = "【Android】[$method] channel requesting"
        debugLog(Category.CHANNEL, message)
        forward(Level.DEBUG, Category.CHANNEL, message)
    }

    fun channelSuccess(method: String) {
        val message = "【Android】[$method] channel success"
        debugLog(Category.CHANNEL, message)
        forward(Level.DEBUG, Category.CHANNEL, message)
    }

    fun channelFailure(method: String, error: String) {
        val message = "【Android】[$method] channel failure: $error"
        debugLog(Category.CHANNEL, message)
        forward(Level.DEBUG, Category.CHANNEL, message)
    }

    // MARK: - Private

    private fun debugLog(category: Category, message: String) {
        if (BuildConfig.DEBUG) {
            Log.d(category.tag, message)
        }
    }

    // TODO: Firebase導入時に FirebaseLogDestination を実装し destinations に追加する
    private fun forward(level: Level, category: Category, message: String) {
        destinations.forEach { it.log(level, category, message) }
    }
}
